src/db/mysql/models.rs

```diff
@@ -0,0 +1,352 @@
+                .id
+        };
+        Ok(id)
+    }
+
+    pub fn get_collections_sync(
+        &self,
+        params: &params::GetCollections,
+    ) -> Result<results::GetCollections> {
+        sql_query("SELECT id, modified FROM collections WHERE user_id = ?")
+            .bind::<Integer, _>(params.user_id as i32)
+            .load::<CollectionsResult>(&self.conn)?
+            .into_iter()
+            .map(|cr| {
+                self.get_collection_name(cr.id)
+                    .map(|name| (name, cr.modified))
+            })
+            .collect()
+    }
+
+    pub fn create_collection_sync(&self, name: &str) -> Result<i32> {
+        let collection_id = self.conn.transaction(|| {
+            sql_query("INSERT INTO collections (name, modified) VALUES (?, ?)")
+                .bind::<Text, _>(name)
+                .bind::<BigInt, _>(ms_since_epoch())
+                .execute(&self.conn)?;
+            collections::table.select(last_insert_id).first(&self.conn)
+        })?;
+        Ok(collection_id)
+    }
+
+    fn get_collection_name(&self, id: i32) -> Result<String> {
+        // XXX: python caches collection names/ids in memory as they're added
+        let name = if let Some(name) = get_std_collection_name(id) {
+            name.to_owned()
+        } else {
+            sql_query("SELECT name FROM collections where id = ?")
+                .bind::<Integer, _>(&id)
+                .get_result::<NameResult>(&self.conn)?
+                .name
+        };
+        Ok(name)
+    }
+
+    pub fn put_bso_sync(&self, bso: &params::PutBso) -> Result<results::PutBso> {
+        /*
+        if bso.payload.is_none() && bso.sortindex.is_none() && bso.ttl.is_none() {
+            // XXX: go returns an error here (ErrNothingToDo), and is treated
+            // as other errors
+            return Ok(());
+        }
+        */
+
+        // XXX: consider mysql ON DUPLICATE KEY UPDATE?
+        self.conn.transaction(|| {
+            let q = r#"
+                SELECT 1 as count FROM bso
+                WHERE user_id = ? AND collection_id = ? AND id = ?
+            "#;
+            let exists = sql_query(q)
+                .bind::<Integer, _>(bso.user_id as i32) // XXX:
+                .bind::<Integer, _>(&bso.collection_id)
+                .bind::<Text, _>(&bso.id)
+                .get_result::<Count>(&self.conn)
+                .optional()?
+                .is_some();
+
+            if exists {
+                update(bso::table)
+                    .filter(bso::user_id.eq(bso.user_id as i32)) // XXX:
+                    .filter(bso::collection_id.eq(&bso.collection_id))
+                    .filter(bso::id.eq(&bso.id))
+                    .set(put_bso_as_changeset(&bso))
+                    .execute(&self.conn)?;
+            } else {
+                let payload = bso.payload.as_ref().map(Deref::deref).unwrap_or_default();
+                let sortindex = bso.sortindex.unwrap_or_default();
```

Comment:
It's not obvious to me whether this defaults sortindex to `NULL` or to `0`; I believe `NULL` is the correct behaviour.

Reply:
good catch, this is leftover from the port from go which instead defaulted to 0.